http_call.go

```diff
@@ -24,6 +25,102 @@ import (
 	"sync"
 )
 
+type unaryHTTPCall struct {
+	httpClient       HTTPClient
+	onRequestSend    func(*http.Request)
+	validateResponse func(*http.Response) *Error
+
+	request  *http.Request
+	response *http.Response // nil until the request has been sent
+}
+
+func newUnaryHTTPCall(
+	ctx context.Context,
+	httpClient HTTPClient,
+	url *url.URL,
+	header http.Header,
+) *unaryHTTPCall {
+	// ensure we make a copy of the url before we pass along to the
+	// Request. This ensures if a transport out of our control wants
+	// to mutate the req.URL, we don't feel the effects of it.
+	url = cloneURL(url)
+
+	// This is mirroring what http.NewRequestContext did, but
+	// using an already parsed url.URL object, rather than a string
+	// and parsing it again. This is a bit funny with HTTP/1.1
+	// explicitly, but this is logic copied over from
+	// NewRequestContext and doesn't effect the actual version
+	// being transmitted.
+	request := (&http.Request{
+		Method:     http.MethodPost,
+		URL:        url,
+		Header:     header,
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		Body:       nil, // Set in Do.
+		Host:       url.Host,
+	}).WithContext(ctx)
+	return &unaryHTTPCall{
+		httpClient: httpClient,
+		request:    request,
+	}
+}
+
+// Do sends the request and waits for the response.
+func (u *unaryHTTPCall) Do(body *bytes.Buffer) *Error {
+	// Body can be nil for GET requests.
+	if body != nil {
+		u.request.Body = io.NopCloser(body)
+		u.request.ContentLength = int64(body.Len())
+
+		// We need to set the GetBody function so that net/http can re-send the
+		// request if required.
+		bodyBytes := body.Bytes()
+		u.request.GetBody = func() (io.ReadCloser, error) {
+			return io.NopCloser(bytes.NewReader(bodyBytes)), nil
+		}
+	}
+	// Promote the header Host to the request object.
+	if host := u.request.Header.Get(headerHost); len(host) > 0 {
+		u.request.Host = host
+	}
+	if u.onRequestSend != nil {
+		u.onRequestSend(u.request)
+	}
+	// Complete the request.
+	response, err := u.httpClient.Do(u.request)
+	if err != nil {
+		err = wrapIfContextError(err)
+		err = wrapIfLikelyH2CNotConfiguredError(u.request, err)
+		err = wrapIfLikelyWithGRPCNotUsedError(err)
+		err = wrapIfRSTError(err)
+		if cerr, ok := asError(err); ok {
+			return cerr
+		}
+		return NewError(CodeUnavailable, err)
+	}
+	u.response = response
+	return u.validateResponse(response)
```

Comment:
If the protocol client fails to call `SetValidateResponse`, this will panic. Maybe a nil guard? Or maybe this should just return the response and let the caller validate it instead of this callback approach.

Reply:
Current behaviour is to panic if not provided. I think it would be nice if the `newDuplexCall` took it as an arg, but can also check for the function. Resolving for now.